packages/math/base-elements.lua

```diff
@@ -744,24 +752,37 @@ elements.terminal = pl.class({
 elements.space = pl.class({
   _base = elements.terminal,
   _type = "Space",
-  __tostring = function(self)
-    return "space{w = "..self.width..", h = "..self.height..", d = "..self.depth.."}"
+  __tostring = function (self)
+    return self._type .. "(width=" .. tostring(self.width) ..
+      ", height=" .. tostring(self.height) ..
+      ", depth=" .. tostring(self.depth) .. ")"
   end,
   _init = function(self, width, height, depth)
     elements.terminal._init(self)
     self.width = width
     self.height = height
     self.depth = depth
   end,
-  shape = function (_) end,
+  shape = function (self)
+    self.width = self.width:absolute() * self:getScaleDown()
```

Comment:
Do you confirm the `absolute()` is required here for some reason? For operations on length, these need to be absolute indeed, but t's a bit dangerous to absolutize "in the wild" without e.g. using the current font options at the time the node was pushed, but not yet shaped. Just asking because we already have, IMHO, plenty of cases due to "too early or too late" absolutizing of relative units (e.g. "1em").

Reply:
Without the call to `absolute()`, the parent stackbox [will try to add relative and absolute lengths together](https://github.com/sile-typesetter/sile/blob/10b333f24f142673ee0dc2bfc23a75b4aecea253/packages/math/base-elements.lua#L462), which causes an error. Would there be a better way to handle this?

Reply:
The issue is less the "way" to handle this as the "when". Is when this function gets called (and the scope it is called from) the right time in the process and does it have the right context to do so?

Reply:
Unless mistaken, `self.options` (inherited from the mbox) contain the options that were in effect when the node was created (which could be a different scope from that where it is pushed, but anyway). So could look as something such as:
```
  SILE.settings:temporarily(function ()
    SILE.settings:set("font.size", self.options.size)
    xxx = xxx:absolute()
  end)
```
(But I don't know that part of the code well, just seeing too that other places refer to `self.options`, e.g. to evaluate the nnodes, etc.)

Reply:
> which could be a different scope from that where it is pushed, but anyway
As far as I checked, the only things that are created and pushed at different place are lrskips, parskips, parindents, etc. Hence my "but anyway" - usual box contents are usually built and pushed to the typesetter queue in a row.

Reply:
(The mess not absolutizing in right context lead to things such as #1361)

Reply:
Thanks for the pointers. It does seem to me like this is the right place to do the conversion, since other `shape()` functions also use the current font size in their calculations. However it hadn’t occurred to me that the math font size is a separate setting from the main text font size. So your solution seems correct @Omikhleia, if we also take into account the scale down factor from `getScaleDown()`.

Reply:
I think a cleaner way would be to define a math unit as in TeX, that is relative to the math font size, while the em stays relative to the main font size. Distinguishing between the two will be especially important when we later add non-math text inside math formulas.

Reply:
I implemented this strategy in the latest commit. I kept the original commit that implemented relative spaces, to retain the context of our discussion for now.

Reply:
While working on this, I noticed that [`getMathMetrics()`](https://github.com/sile-typesetter/sile/blob/5a6de0ca3a71163b56b5bf1ebb0eac1deb92c420/packages/math/base-elements.lua#L129) caches its result, but the cache doesn’t seem to be invalidated when the font family or size is changed. Am I missing something here @OlivierNicole?

Reply:
No, you are right, that was one of the things I was supposed to fix “later”…

Reply:
I added a fix for this in the latest commit.